hw/ip/hmac/doc/programmers_guide.md

```diff
@@ -88,14 +88,17 @@ To switch from one message stream to another, set the `CMD.hash_stop` bit, wait
 
 Here is an example usage pattern of this feature:
 1. Start processing message stream A by configuring HMAC and then setting the `CMD.hash_start` bit.
-2. Write an arbitrary number of message blocks to HMAC's `MSG_FIFO`.
-3. Stop HMAC by setting the `CMD.hash_stop` bit and wait for the `hmac_done` interrupt (or poll the interrupt status register).
-4. Save the context by reading the `DIGEST_0`..`15` and `MSG_LENGTH_`{`LOWER`,`UPPER`} registers. (The values in the `CFG` register must also be preserved, but that is purely SW-controlled so doesn't need to be read from HW.)
-5. Repeat steps 1-4 for message stream B.
-6. Restore the context of message stream A by writing the `CFG`, `DIGEST_0`..`15`, and `MSG_LENGTH_`{`LOWER`,`UPPER`} registers.
-7. Continue processing message stream A by setting the `CMD.hash_continue` bit.
-8. Write an arbitrary number of message blocks to HMAC's `MSG_FIFO`.
-9. Continue this with as many message blocks and parallel message streams as needed. The final hash for any message stream can be obtained at any time (no need for complete blocks) by setting `CMD.hash_process` and waiting for the `hmac_done` interrupt / status bit, finally reading the digest from the `DIGEST` registers.
+1. Write an arbitrary number of message blocks to HMAC's `MSG_FIFO`.
+1. Stop HMAC by setting the `CMD.hash_stop` bit and wait for the `hmac_done` interrupt (or poll the interrupt status register).
+1. Save the context by reading the `DIGEST_0`..`15` and `MSG_LENGTH_`{`LOWER`,`UPPER`} registers. (The values in the `CFG` register must also be preserved, but that is purely SW-controlled so doesn't need to be read from HW.)
```

Comment:
I think the `key` registers should also be preserved as values can be changed

Reply:
I think you're right, because the key not only gets hashed *before* the message (in the first round) but also *after* the message (in the second round) before the final digest is computed.